nexus/src/app/sagas/region_snapshot_replacement_step.rs

```diff
@@ -0,0 +1,603 @@
+            // - unwind so that this saga can run again.
+            return Err(ActionError::action_failed(format!(
+                "vmm {} propolis not in a state to receive request",
+                vmm.id,
+            )));
+        }
+    }
+
+    let new_volume_vcr = match osagactx
+        .datastore()
+        .volume_get(params.request.volume_id)
+        .await
+        .map_err(ActionError::action_failed)?
+    {
+        Some(volume) => volume.data().to_string(),
+
+        None => {
+            return Err(ActionError::action_failed(Error::internal_error(
+                "new volume is gone!",
+            )));
+        }
+    };
+
+    let instance_lookup =
+        LookupPath::new(&opctx, &osagactx.datastore()).instance_id(instance_id);
+
+    let (vmm, client) = osagactx
+        .nexus()
+        .propolis_client_for_instance(
+            &opctx,
+            &instance_lookup,
+            authz::Action::Modify,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    info!(
+        log,
+        "sending replacement request for disk volume to propolis";
+        "request id" => %params.request.id,
+        "volume id" => %params.request.volume_id,
+        "disk id" => ?disk.id(),
+        "instance id" => ?instance_id,
+        "vmm id" => ?vmm.id,
+    );
+
+    let result = client
+        .instance_issue_crucible_vcr_request()
+        .id(disk.id())
+        .body(propolis_client::types::InstanceVcrReplace {
+            name: disk.name().to_string(),
+            vcr_json: new_volume_vcr,
+        })
+        .send()
+        .await
+        .map_err(|e| match e {
+            propolis_client::Error::ErrorResponse(rv) => {
+                ActionError::action_failed(rv.message.clone())
+            }
+
+            _ => ActionError::action_failed(format!(
+                "unexpected failure during \
+                        `instance_issue_crucible_vcr_request`: {e}",
+            )),
+        })?;
+
+    let replace_result = result.into_inner();
+
+    info!(
+        log,
+        "saw replace result {replace_result:?}";
+        "request id" => %params.request.id,
+        "volume id" => %params.request.volume_id,
+        "disk id" => ?disk.id(),
+        "instance id" => ?instance_id,
+        "vmm id" => ?vmm.id,
+    );
+
+    match &replace_result {
+        ReplaceResult::Started => {
+            // This saga's call just started the replacement
+        }
+
+        ReplaceResult::StartedAlready => {
+            // A previous run of this saga (or saga node) started the
+            // replacement
+        }
+
+        ReplaceResult::CompletedAlready => {
+            // It's done! We see this if the same propolis that received the
+            // original replace request started and finished the replacement.
+        }
+
+        ReplaceResult::VcrMatches => {
+            // This propolis booted with the updated VCR
```

Comment:
This saga takes a lock on the region snapshot replacement request, so it wouldn't be another dueling saga in this case. One way to hit this case is if this function crashes before it can be recorded in crdb that it successfully happened. When the executor picks this saga up to run again, it will re-execute this node, re-sending the replacement request which will return VcrMatches (due to the previously sent replacement request).